### ISSUE TITLE
task_sentiment_albert 训练集准确率50%，测试集准确率94%

### ISSUE THREAD

**Issue by hurun** [bojone/bert4keras#267]
提问时请尽可能提供如下信息：

### 基本信息
- 你使用的**操作系统**: linux
- 你使用的**Python**版本: 3.6.9
- 你使用的**Tensorflow**版本: 2.3.1
- 你使用的**Keras**版本: 2.4.3
- 你使用的**bert4keras**版本: 0.9.1
- 你使用纯**keras**还是**tf.keras**: keras
- 你加载的**预训练模型**:albert

### 核心代码
```
output = Lambda(lambda x: x[:, 0], name='CLS-token')(bert.model.output)
output = Dense(
    units=num_classes,
    activation='softmax',
    kernel_initializer=bert.initializer
)(output)

model.fit(
        train_generator.forfit(),
        steps_per_epoch=len(train_generator),
        epochs=10,
        callbacks=[evaluator]
    )
# 请在此处贴上你的核心代码。
# 请尽量只保留关键部分，不要无脑贴全部代码。
```

### 输出信息
```shell
Epoch 1/10
528/528 [==============================] - ETA: 0s - loss: 0.2835 - accuracy: 0.4854val_acc: 0.92563, best_val_acc: 0.92563, test_acc: 0.90715

528/528 [==============================] - 107s 202ms/step - loss: 0.2835 - accuracy: 0.4854
Epoch 2/10
528/528 [==============================] - ETA: 0s - loss: 0.1550 - accuracy: 0.4948val_acc: 0.92894, best_val_acc: 0.92894, test_acc: 0.92042

528/528 [==============================] - 105s 199ms/step - loss: 0.1550 - accuracy: 0.4948
Epoch 3/10
528/528 [==============================] - ETA: 0s - loss: 0.0988 - accuracy: 0.4939val_acc: 0.93747, best_val_acc: 0.93747, test_acc: 0.93131

528/528 [==============================] - 105s 200ms/step - loss: 0.0988 - accuracy: 0.4939
Epoch 4/10
528/528 [==============================] - ETA: 0s - loss: 0.0725 - accuracy: 0.4944val_acc: 0.93605, best_val_acc: 0.93747, test_acc: 0.93273

528/528 [==============================] - 105s 199ms/step - loss: 0.0725 - accuracy: 0.4944
Epoch 5/10
528/528 [==============================] - ETA: 0s - loss: 0.0494 - accuracy: 0.4938val_acc: 0.93652, best_val_acc: 0.93747, test_acc: 0.93131

528/528 [==============================] - 105s 199ms/step - loss: 0.0494 - accuracy: 0.4938
Epoch 6/10
528/528 [==============================] - ETA: 0s - loss: 0.0454 - accuracy: 0.4930val_acc: 0.94126, best_val_acc: 0.94126, test_acc: 0.93700

528/528 [==============================] - 105s 199ms/step - loss: 0.0454 - accuracy: 0.4930
Epoch 7/10
528/528 [==============================] - ETA: 0s - loss: 0.0400 - accuracy: 0.4932val_acc: 0.93794, best_val_acc: 0.94126, test_acc: 0.93415

528/528 [==============================] - 105s 198ms/step - loss: 0.0400 - accuracy: 0.4932
Epoch 8/10
528/528 [==============================] - ETA: 0s - loss: 0.0267 - accuracy: 0.4913val_acc: 0.94268, best_val_acc: 0.94268, test_acc: 0.92800

528/528 [==============================] - 105s 199ms/step - loss: 0.0267 - accuracy: 0.4913
Epoch 9/10
528/528 [==============================] - ETA: 0s - loss: 0.0268 - accuracy: 0.4915val_acc: 0.94268, best_val_acc: 0.94268, test_acc: 0.93842

528/528 [==============================] - 105s 198ms/step - loss: 0.0268 - accuracy: 0.4915
Epoch 10/10
528/528 [==============================] - ETA: 0s - loss: 0.0234 - accuracy: 0.4914val_acc: 0.94268, best_val_acc: 0.94268, test_acc: 0.93036

528/528 [==============================] - 105s 199ms/step - loss: 0.0234 - accuracy: 0.4914
final test acc: 0.927996
```

### 自我尝试
不管什么问题，请先尝试自行解决，“万般努力”之下仍然无法解决再来提问。此处请贴上你的努力过程。
尝试改变softmax 为sigmoid ，units=1, loss='binary_crossentropy'，能把训练集准确率提示至1.00，但测试集效果降低为51%

想请教可以改变代码哪个部分能输出正确结果

**Comment by bojone** [comment]
目测是keras版本和tf版本太新吧。

可以尝试：
- `metrics=['accuracy']`改为`metrics=['sparse_categorical_accuracy']`；
- 换用keras 2.3.1 + tf <= 2.2
- 换用tf.keras（导入bert4keras前传入环境变量`TF_KERAS=1`）

**Comment by hurun** [comment]
统一改成了sigmoid的方式，修改了evaluate评价验证集和测试集的方法，能成功跑出结果，准确率93-95%

output = Lambda(lambda x: x[:, 0], name='CLS-token')(bert.model.output)
# output = Dense(256, 'relu')(output)
output = Dense(32, 'relu')(output)
output = Dense(
    units=1,
    activation='sigmoid',
    kernel_initializer=bert.initializer
)(output)

model = keras.models.Model(bert.model.input, output)
# model.summary()

# 派生为带分段线性学习率的优化器。
# 其中name参数可选，但最好填入，以区分不同的派生优化器。
AdamLR = extend_with_piecewise_linear_lr(Adam, name='AdamLR')

model.compile(
    loss='binary_crossentropy',
    optimizer=Adam(1e-5),  # 用足够小的学习率
    # optimizer=AdamLR(learning_rate=1e-4, lr_schedule={
    #     1000: 1,
    #     2000: 0.1
    # }),
    metrics=['accuracy'],
)



def evaluate(data):
    total, right = 0., 0.
    print_flag = 1
    for x_true, y_true in data:
        y_pred = np.where(model.predict(x_true) > 0.5, 1, 0)
        y_pred = y_pred.reshape(1, -1)[0]
        y_true = y_true[:, 0]
        # if print_flag:
        #   # print('model',model.predict(x_true))
        #   print('y_true', y_true[:3])
        #   print('y_pred', y_pred[:3])
        #   print_flag=0
        total += len(y_true)
        right += (y_true == y_pred).sum()
    return right / total
